File: its-my-training-day/training-grounds/leetcode-top-75/questions/dynamic-programming/dp-lis/longest-increasing-subsequence/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

int longestIncreasingSubsequence(const std::vector<int>& nums) {
    if (nums.empty()) return 0;
    std::vector<int> tails(nums.size(), 0);
    int size = 0;
    for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
        int val = nums[i];
        int left = 0;
        int right = size;
        while (left < right) {
            int mid = (left + right) / 2;
            if (tails[mid] < val) left = mid + 1;
            else right = mid;
        }
        tails[left] = val;
        if (left == size) size += 1;
    }
    return size;
}

File: its-my-training-day/training-grounds/leetcode-top-75/questions/dynamic-programming/dp-lis/longest-increasing-subsequence/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def longest_increasing_subsequence(nums: List[int]) -> int:
    if not nums:
        return 0
    tails = [0] * len(nums)
    size = 0
    i = 0
    while i < len(nums):
        val = nums[i]
        left = 0
        right = size
        while left < right:
            mid = (left + right) // 2
            if tails[mid] < val:
                left = mid + 1
            else:
                right = mid
        tails[left] = val
        if left == size:
            size += 1
        i += 1
    return size
